### PR TITLE
Remove C++ density matrix machines + jax workarounds for R->C machines

### PR DESCRIPTION
**Pull request by PhilipVinc**
Notable changes:
 - new `outdtype` in Jax machine constructor, used to determine how to compute the gradient (because Jax does not like R->C functions).
 - new `vip` implementation in machines, which is now fitted compared to before.
 - added l2 norm to the density matrix `to_array`

### CODE REVIEW

**Comment by PhilipVinc** [comment]
I also removed the PyBind export of MPI from the C module and all it's usages from netket itself.

While I'd also like to remove MPI from the C library, doing so in a non-test breaking way takes more effort because random_utils depends on it, Spin hilbert depends on random_utils, and... well. we could just get rid of C++ hilbert but then MPS machines would not work anymore.

I think, now that the Graph PR has been merged, we could seriously just remove all C++ references and in case add them later...

**Comment by gcarleo** [comment]
Yes thanks! I will do a separate PR transitioning to the new Hilbert and to the new Graph. The only issue is that graph.Lattice is not ready yet, but maybe I can just comment that part in the tests for the moment

File: Test/Machine/test_machine.py [hunk 1]
```diff
@@ -63,6 +64,10 @@ def initializer(rng, shape):
         dtype=complex,
     )
 
+    dm_machines["Jax NDM"] = nk.machine.density_matrix.JaxNdmSpinPhase(
```
Comment (gcarleo) [inline Test/Machine/test_machine.py:67]: I was thinking that maybe we should just call this NdmSpinPhase and remove Jax... ?

File: netket/operator/der_local_values_jax.py [hunk 2]
```diff
@@ -75,7 +74,7 @@ def _local_values_and_grads_notcentered_kernel(pars, vp, mel, v, logpsi):
 
 def _der_local_values_notcentered_impl(op, machine, v, log_vals):
     sections = _np.empty(v.shape[0], dtype=_np.int32)
-    v_primes, mels = op.get_conn_flattened(v, sections, pad=True)
+    v_primes, mels = op.get_conn_flattened(v._value, sections, pad=True)
```
Comment (gcarleo) [inline netket/operator/der_local_values_jax.py:77]: is using _value safe though? This looks like an internal attribute that might not be always consistent?

File: Test/Machine/test_machine.py (hunk 1 above)
Comment (PhilipVinc) [reply Test/Machine/test_machine.py:67]: Since it's really a Jax machine, which is an optional dependency, and maybe in 2025 when I'll be old i'll write a Torch version, I'd leave this as is for now.

File: netket/operator/der_local_values_jax.py (hunk 2 above)
Comment (PhilipVinc) [reply netket/operator/der_local_values_jax.py:77]: `._value` is not stable API according to Jax, but it's used internally for the exact same purpose in the Jax-supported wrappers of DLPack. I doubt there is another way to do this without copying.

Comment (PhilipVinc) [reply netket/operator/der_local_values_jax.py:77]: I'd keep this like that for the time being, and if Jax stabilises the api we can upgrade too.
